### PR TITLE
Issue #2719479: Add field formatter for variations field to render av…

### PR DESCRIPTION
**Pull request by mglaman**
…ailable attribute options

### CODE REVIEW

File: modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+   */
+  protected function getAttributeItemList(FieldItemListInterface $variations, ProductAttributeInterface $attribute) {
+    $build = [
+      '#theme' => 'item_list',
+      '#title' => $attribute->label(),
+      '#items' => [],
+    ];
+
+    /** @var \Drupal\Core\Render\RendererInterface $renderer */
+    $renderer = \Drupal::service('renderer');
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php]: We have the renderer injected, let's use it :)

File: modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+ *   label = @Translation("Product attributes overview"),
+ *   field_types = {
+ *     "entity_reference",
+ *   },
+ * )
+ */
+class ProductAttributesOverview extends FormatterBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The entity manager.
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php]: The entity type manager.

File: modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+  /**
+   * The renderer.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * Constructs a FormatterBase object.
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php]: Constructs a new ProductAttributesOverview object.

**Comment by bojanz** [comment]
We want to make the view mode configurable, and make sure that linking the rendered attribute to the product works. Also, a test.

File: modules/product/config/schema/commerce_product.schema.yml [hunk 4]
```diff
@@ -59,6 +59,14 @@ field.formatter.settings.commerce_add_to_cart:
       type: boolean
       label: 'Whether to attempt to combine line items containing the same product variation'
 
+field.formatter.settings.commerce_product_attributes_overview:
+  type: mapping
+  mapping:
+    attributes:
+      type: sequence
+      label: 'Attributes to render.'
+
```
Comment (bojanz) [inline modules/product/config/schema/commerce_product.schema.yml]: Double newline.

**Comment by mglaman** [comment]
Made display mode configurable and started on test.

@bojanz do we support linking to a specific variation anyways? This doesn't make each attribute a link.

File: modules/product/config/schema/commerce_product.schema.yml [hunk 5]
```diff
@@ -59,6 +59,20 @@ field.formatter.settings.commerce_add_to_cart:
...
+      type: sequence
+      label: 'Attributes to render.'
+      sequence:
+          type: string
+          label: 'Attribute name'
+    view_mode:
+      type: string
+      label: 'View mode'
+
+
```
Comment (bojanz) [inline modules/product/config/schema/commerce_product.schema.yml]: Double newline.

File: modules/product/config/schema/commerce_product.schema.yml [hunk 6]
```diff
@@ -59,6 +59,20 @@ field.formatter.settings.commerce_add_to_cart:
...
       label: 'Whether to attempt to combine line items containing the same product variation'
 
+field.formatter.settings.commerce_product_attributes_overview:
+  type: mapping
+  mapping:
+    attributes:
+      type: sequence
+      label: 'Attributes to render.'
+      sequence:
+          type: string
```
Comment (bojanz) [inline modules/product/config/schema/commerce_product.schema.yml]: Wrong indentation.

File: modules/product/tests/src/Kernel/ProductAttributesOverviewFormatterTest.php [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        'view_mode' => 'default',
+      ],
+    ]);
+    $this->productDefaultDisplay->save();
+
+    $build = $this->productViewBuilder->view($this->product);
+    $this->render($build);
+
+    $this->assertRaw('<div>
+    <div>Variations</div>
```
Comment (bojanz) [inline modules/product/tests/src/Kernel/ProductAttributesOverviewFormatterTest.php]: This sounds really fragile, any change to the markup will break the test. Better to make it less specific by testing for expected pieces (individual colors, labels).

Comment (mglaman) [inline modules/product/tests/src/Kernel/ProductAttributesOverviewFormatterTest.php]: The problem is there are no classes, and field injection is working. So asserting for just a color can also catch the variation titles.

Comment (mglaman) [inline modules/product/tests/src/Kernel/ProductAttributesOverviewFormatterTest.php]: Also attempting any partial of that fails (shrug)

File: modules/product/config/schema/commerce_product.schema.yml [hunk 8]
```diff
@@ -62,6 +62,19 @@ field.formatter.settings.commerce_add_to_cart:
...
 
+field.formatter.settings.commerce_product_attributes_overview:
+  type: mapping
+  mapping:
+    attributes:
+      type: sequence
+      label: 'Attributes to render.'
+      sequence:
+        type: string
+        label: 'Attribute name'
```
Comment (bojanz) [inline modules/product/config/schema/commerce_product.schema.yml]: Attribute ID, not name.

File: modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+   * @param \Drupal\Core\Entity|EntityDisplayRepositoryInterface $entity_display_repository
+   *   The entity display repository.
+   * @param \Drupal\commerce_product\ProductAttributeFieldManagerInterface $attribute_field_manager
+   *   The attribute field manager.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer.
+   */
+  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, $label, $view_mode, array $third_party_settings, EntityTypeManagerInterface $entity_type_manager, EntityDisplayRepositoryInterface $entity_display_repository, ProductAttributeFieldManagerInterface $attribute_field_manager, RendererInterface $renderer) {
+    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $label, $view_mode, $third_party_settings);
+    $this->entityTypeManager = $entity_type_manager;
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php:88]: It's probably more readable to have an empty line after parent::__construct when we have a bunch of lines following.

File: modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $product_attribute_storage = $this->entityTypeManager->getStorage('commerce_product_attribute');
+    $elements = [];
+    foreach (array_filter($this->getSetting('attributes')) as $item) {
+      $attribute = $product_attribute_storage->load($item);
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php]: We should load all product attributes at once, not one by one.

File: modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    $build = [
+      '#theme' => 'item_list',
+      '#title' => $attribute->label(),
+      '#items' => [],
+    ];
+
+    $view_builder = $this->entityTypeManager->getViewBuilder('commerce_product_attribute_value');
+
+    /** @var \Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem $variation */
+    foreach ($variations as $variation) {
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldFormatter/ProductAttributesOverview.php]: These are not really $variations if they're EntityReferenceItem instances. Let's do $variation_items

File: modules/product/tests/src/Kernel/ProductAttributesOverviewFormatterTest.php [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        'view_mode' => 'default',
+      ],
+    ]);
+    $this->productDefaultDisplay->save();
+
+    $build = $this->productViewBuilder->view($this->product);
+    $this->render($build);
+
+    $this->assertFieldByXPath('//h3[text()=\'Color\']');
+    $this->assertFieldByXPath('//ul/li[1]/a/div/div/div[text()=\'Name\']');
```
Comment (bojanz) [inline modules/product/tests/src/Kernel/ProductAttributesOverviewFormatterTest.php:145]: Nice!

**Comment by bojanz** [comment]
Committed in 99885d0dd25b057e6b83825129ece0b33c964ea4.

**Pull request closed without merging**